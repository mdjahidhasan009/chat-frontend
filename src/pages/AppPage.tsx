import { LayoutPage, Page } from "../utils/styles";
import { Outlet, useLocation, useNavigate } from "react-router-dom";
import { UserSidebar} from "../components/sidebars/UserSidebar";
import { useContext, useEffect, useRef } from "react";
import { useDispatch, useSelector } from "react-redux";
import { SocketContext } from "../utils/context/SocketContext";
import { AppDispatch, RootState } from "../store";
import { AcceptFriendRequestResponse, AcceptedCallPayload, FriendRequest, SelectableTheme, CallPayload } from "../utils/types";
import { addFriendRequest, removeFriendRequest } from "../store/friends/friendsSlice";
import { useToast } from "../utils/hooks/useToast";
import { IoMdPersonAdd } from 'react-icons/io';
import { BsFillPersonCheckFill } from 'react-icons/bs';
import { fetchFriendRequestThunk } from "../store/friends/friendsThunk";
import { ThemeProvider } from "styled-components";
import { DarkTheme, LightTheme } from "../utils/themes";
import { AuthContext } from "../utils/context/AuthContext";
import Peer from "peerjs";
import { setActiveConversationId, setCall, setCaller, setConnection, setIsCallInProgress, setIsReceivingCall, setLocalStream, setPeer, setReceiver, setRemoteStream } from "../store/call/callSlice";
import { useVideoCallRejected } from "../utils/hooks/sockets/useVideoCallRejected";
import { useVideoCallHangUp } from "../utils/hooks/sockets/useVideoCallHangUp";
import { useVideoCallAccept } from "../utils/hooks/sockets/useVideoCallAccept";
import { useFriendRequestReceived } from "../utils/hooks/sockets/friend-requests/useFriendRequestReceived";
import { useVideoCall } from "../utils/hooks/sockets/call/useVideoCall";
import { useVoiceCall } from "../utils/hooks/sockets/call/useVoiceCall";
import { useVoiceCallAccepted } from "../utils/hooks/sockets/call/useVoiceCallAccepted";
import { useVoiceCallHangUp } from "../utils/hooks/sockets/call/useVoiceCallHangUp";
import { useVoiceCallRejected } from "../utils/hooks/sockets/call/useVoiceCallRejected";
import { CallReceiveDialog } from "../components/calls/CallReceiveDialog";

export const AppPage = () => {
  const { user } = useContext(AuthContext);
  const { pathname } = useLocation();
  const socket = useContext(SocketContext);
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();
  const { peer, call, isReceivingCall, caller, connection, callType } = useSelector(
    (state: RootState) => state.call
  );
  const { info } = useToast({ theme: 'dark' });
  const { theme } = useSelector((state: RootState) => state.settings);
  const storageTheme = localStorage.getItem('theme') as SelectableTheme;

  useEffect(() => {
    dispatch(fetchFriendRequestThunk());
  }, [dispatch]);

  // useEffect(() => {
  //   if(!user) return;
  //   const newPeer = new Peer(user.peer.id, {
  //     config: {
  //       iceServers: [
  //         {
  //           url: 'stun:stun.l.google.com:19302',
  //         },
  //         {
  //           url: 'stun:stun1.l.google.com:19302',
  //         },
  //       ],
  //     },
  //   });
  //
  //   peer?.on('error', (err) => {
  //     console.error('PeerJS Error:', err);
  //   });
  //
  //   peer?.on('disconnected', () => {
  //     console.warn('Peer disconnected. Attempting reconnect...');
  //     peer.reconnect(); // Attempt to reconnect
  //   });
  //
  //   peer?.on('open', (id) => {
  //     console.log('Peer connection established with ID:', id);
  //   });
  //   // debugger
  //   dispatch(setPeer(newPeer));
  // }, []);

  useFriendRequestReceived();
  useVideoCall();

  useEffect(() => {
    socket.on('onFriendRequestCancelled', (payload: FriendRequest) => {
      dispatch(removeFriendRequest(payload));
    });

    socket.on('onFriendRequestAccepted', (payload: AcceptFriendRequestResponse) => {
      dispatch(removeFriendRequest(payload.friendRequest));
      socket.emit('getOnlineFriends');
      info(`${payload.friendRequest.receiver.firstName} accepted your friend request`, {
          position: 'bottom-left',
          icon: BsFillPersonCheckFill,
          onClick: () => navigate('/friends'),
        }
      );
    });

    socket.on('onFriendRequestRejected', (payload: FriendRequest) => {
      dispatch(removeFriendRequest(payload));
    });

    return () => {
      socket.off('onFriendRequestCancelled');
      socket.off('onFriendRequestRejected');
      socket.off('onFriendRequestReceived');
      socket.off('onFriendRequestAccepted');
    };
  }, [socket, isReceivingCall]);

    /**
   * This useEffect hook is for the user who is receiving the call.
   * So we must dispatch the appropriate actions to set the state
   * for the user receiving the call.
   *
   * The user who is calling will have its own instance of MediaConnection/Call
   */
  // useEffect(() => {
  //   if(!peer) return;
  //   console.log('peer')
  //   // debugger
  //   debugger
  //
  //   peer.on('call', async (incomingCall) => {
  //     const constraints = { video: callType === 'video', audio: true };
  //     // debugger
  //     debugger
  //     const stream = await navigator.mediaDevices.getUserMedia(constraints);
  //     incomingCall.answer(stream);
  //     dispatch(setLocalStream(stream));
  //     dispatch(setCall(incomingCall));
  //   });
  //   return () => {
  //     peer.off('call');
  //   };
  // }, [peer, callType, dispatch]);

  // useEffect(() => {
  //   if (!call) return;
  //   call.on('stream', (remoteStream) =>
  //     dispatch(setRemoteStream(remoteStream))
  //   );
  //   call.on('close', () => console.log('call was closed'));
  //   return () => {
  //     call.off('stream');
  //     call.off('close');
  //   };
  // }, [call]);

  console.log('AppPage.tsx')
  useVideoCallAccept();
  useVideoCallRejected();
  useVideoCallHangUp();
  useVoiceCall();
  useVoiceCallAccepted();
  useVoiceCallHangUp();
  useVoiceCallRejected();

  // useEffect(() => {
  //   if(connection) {
  //     connection.on('open', () => {
  //       console.log('connection opened');
  //     });
  //     connection.on('error', () => {
  //       console.log('connection error');
  //     });
  //     connection.on('data', (data) => {
  //       console.log('data received', data);
  //     })
  //     connection.on('close', () => {
  //       console.log('connection closed');
  //     });
  //   };
  //   return () => {
  //     connection?.off('open');
  //     connection?.off('error');
  //     connection?.off('data');
  //   };
  // }, [connection]);

  return (
    <ThemeProvider
      theme={
        storageTheme
          ? storageTheme === 'dark'
            ? DarkTheme
            : LightTheme
          : theme === 'dark'
          ? DarkTheme
          : LightTheme
      }
    >
    {isReceivingCall && caller && <CallReceiveDialog />}
    <LayoutPage>
      <UserSidebar />
      <Outlet />
    </LayoutPage>
  </ThemeProvider>
  )
}