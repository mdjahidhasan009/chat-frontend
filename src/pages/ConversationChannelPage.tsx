import { useContext, useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { MessagePanel } from '../components/messages/MessagePanel';
import { getConversationMessages } from '../utils/api';
import { AuthContext } from '../utils/context/AuthContext';
import { SocketContext } from '../utils/context/SocketContext';
import { ConversationChannelPageStyle } from '../utils/styles';
import { MessageEventPayload, MessageType } from '../utils/types';

export const ConversationChannelPage = () => {
  const { user } = useContext(AuthContext);
  const socket = useContext(SocketContext);
  const [messages, setMessages] = useState<MessageType[]>([]);
  const { id } = useParams();

  useEffect(() => {
    const conversationId = parseInt(id!);
    getConversationMessages(conversationId)
      .then(({ data }) => {
        setMessages(data);
      })
      .catch((err) => console.log(err));
  }, [id]);

  useEffect(() => {
    socket.on('connected', () => console.log('Connected'));
    socket.on('onMessage', (payload: MessageEventPayload) => {
      console.log('Message Received');
      const { conversation, ...message } = payload;
      setMessages((prev) => [message, ...prev]);
    });
    return () => {
      socket.off('connected');
      socket.off('onMessage');
    };
  }, []);

  return (
    <ConversationChannelPageStyle>
      <MessagePanel messages={messages}></MessagePanel>
    </ConversationChannelPageStyle>
  );
};