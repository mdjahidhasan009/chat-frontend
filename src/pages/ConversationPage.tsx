import {Outlet, useParams} from "react-router-dom";
import { Page } from '../utils/styles';
import { ConversationSidebar } from "../components/conversation/ConversationSidebar";
import { ConversationPanel } from "../components/conversation/ConversationPanel";

export const ConversationPage = () => {
    const { id } = useParams();
    return <Page>
        <ConversationSidebar />
        {!id && <ConversationPanel />}
        <Outlet />
    </Page>
}