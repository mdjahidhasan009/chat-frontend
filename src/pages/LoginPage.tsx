import { Page } from "../utils/styles";
import { LoginForm } from "../components/forms/login";

export const LoginPage = () => {
    return (
        <Page display="flex" justifyContent="center" alignItems="center">
            <LoginForm />
        </Page>
    )
};