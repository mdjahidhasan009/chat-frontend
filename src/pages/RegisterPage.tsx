import { RegisterForm } from '../components/forms/RegisterForm';
import { Page } from '../utils/styles';
export const RegisterPage = () => {
    return (
        <Page display="flex" justifyContent="center" alignItems="center">
            <RegisterForm />
        </Page>
    )

}