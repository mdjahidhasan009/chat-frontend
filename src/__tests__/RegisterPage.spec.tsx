import { render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom'
import { RegisterPage } from '../pages/RegisterPage';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

describe('RegisterPage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render form field', () => {
    const view = render(
      <Router>
        <RegisterPage />
      </Router>
    );

    //previous snapsort create error so before run this test remove previous snapshot(__snapshots__ folder)
    expect(view).toMatchSnapshot();
  });

  it('should display all errors when submitting with all empty fields', async() => {
    render(
      <Router>
        <RegisterPage />
      </Router>
    );

    const submitButton = screen.getByRole('button', { name: 'Create My Account' });
    submitButton.click();
    
    const usernameError = await screen.findByText('Username is Required');
    const firstNameError = await screen.findByText('First Name is Required');
    const lastNameError = await screen.findByText('Last Name is Required');
    const passwordError = await screen.findByText('Password is Required');
    await waitFor(() => {
      expect(usernameError).toBeInTheDocument();
    });
    expect(firstNameError).toBeInTheDocument();
    expect(lastNameError).toBeInTheDocument();
    expect(passwordError).toBeInTheDocument();
  });

  it('should submit empty username field then remove error after typing and leaving focus', async() => {
    render(
      <Router>
        <RegisterPage />
      </Router>
    );

    const submitButton = screen.getByRole('button');
    submitButton.click();
    const usernameError = await screen.findByText('Username is Required');
    await waitFor(() => {
      expect(usernameError).toBeInTheDocument();
    });
    const usernameField = screen.getByLabelText('Username');
    const firstNameField = screen.getByLabelText('First Name');
    expect(usernameField).toBeInTheDocument();
    expect(firstNameField).toBeInTheDocument();
    userEvent.type(usernameField, 'helloworld');
    userEvent.click(firstNameField);
    await waitForElementToBeRemoved(() => screen.queryByText('Username is Required'));
  });
})