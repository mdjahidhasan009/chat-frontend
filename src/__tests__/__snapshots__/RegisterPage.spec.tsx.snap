// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPage should render form field 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-iBYQkv dKrEZN"
        display="flex"
      >
        <form
          class="form"
        >
          <div
            class="sc-gswNZR kPWjWl"
          >
            <div
              class="sc-hLBbgP dMMVby"
            >
              <label
                class="sc-eDvSVe fHxpEk"
                for="username"
              >
                Username
              </label>
            </div>
            <input
              class="sc-bcXHqe iLzThh"
              id="username"
              name="username"
              type="text"
            />
          </div>
          <section
            class="nameFieldRow"
          >
            <div
              class="sc-gswNZR kPWjWl"
            >
              <div
                class="sc-hLBbgP dMMVby"
              >
                <label
                  class="sc-eDvSVe fHxpEk"
                  for="firstName"
                >
                  First Name
                </label>
              </div>
              <input
                class="sc-bcXHqe iLzThh"
                id="firstName"
                name="firstName"
                type="text"
              />
            </div>
            <div
              class="sc-gswNZR kPWjWl"
            >
              <div
                class="sc-hLBbgP dMMVby"
              >
                <label
                  class="sc-eDvSVe fHxpEk"
                  for="lastName"
                >
                  Last Name
                </label>
              </div>
              <input
                class="sc-bcXHqe iLzThh"
                id="lastName"
                name="lastName"
                type="text"
              />
            </div>
          </section>
          <div
            class="sc-gswNZR kPWjWl"
          >
            <div
              class="sc-hLBbgP dMMVby"
            >
              <label
                class="sc-eDvSVe fHxpEk"
                for="password"
              >
                Password
              </label>
            </div>
            <div
              class="passwordContainer"
            >
              <input
                class="sc-bcXHqe iLzThh"
                id="password"
                name="password"
                type="password"
              />
              <svg
                cursor="pointer"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M396 512a112 112 0 1 0 224 0 112 112 0 1 0-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 0 0 0 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
                />
              </svg>
            </div>
          </div>
          <button
            class="sc-gKPRtg fFWYtE button"
          >
            Create My Account
          </button>
          <div
            class="footerText"
          >
            <span>
              Already have an account? 
            </span>
            <a
              href="/login"
            >
              <span>
                Login
              </span>
            </a>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-iBYQkv dKrEZN"
      display="flex"
    >
      <form
        class="form"
      >
        <div
          class="sc-gswNZR kPWjWl"
        >
          <div
            class="sc-hLBbgP dMMVby"
          >
            <label
              class="sc-eDvSVe fHxpEk"
              for="username"
            >
              Username
            </label>
          </div>
          <input
            class="sc-bcXHqe iLzThh"
            id="username"
            name="username"
            type="text"
          />
        </div>
        <section
          class="nameFieldRow"
        >
          <div
            class="sc-gswNZR kPWjWl"
          >
            <div
              class="sc-hLBbgP dMMVby"
            >
              <label
                class="sc-eDvSVe fHxpEk"
                for="firstName"
              >
                First Name
              </label>
            </div>
            <input
              class="sc-bcXHqe iLzThh"
              id="firstName"
              name="firstName"
              type="text"
            />
          </div>
          <div
            class="sc-gswNZR kPWjWl"
          >
            <div
              class="sc-hLBbgP dMMVby"
            >
              <label
                class="sc-eDvSVe fHxpEk"
                for="lastName"
              >
                Last Name
              </label>
            </div>
            <input
              class="sc-bcXHqe iLzThh"
              id="lastName"
              name="lastName"
              type="text"
            />
          </div>
        </section>
        <div
          class="sc-gswNZR kPWjWl"
        >
          <div
            class="sc-hLBbgP dMMVby"
          >
            <label
              class="sc-eDvSVe fHxpEk"
              for="password"
            >
              Password
            </label>
          </div>
          <div
            class="passwordContainer"
          >
            <input
              class="sc-bcXHqe iLzThh"
              id="password"
              name="password"
              type="password"
            />
            <svg
              cursor="pointer"
              fill="currentColor"
              height="24"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 1024 1024"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M396 512a112 112 0 1 0 224 0 112 112 0 1 0-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 0 0 0 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
              />
            </svg>
          </div>
        </div>
        <button
          class="sc-gKPRtg fFWYtE button"
        >
          Create My Account
        </button>
        <div
          class="footerText"
        >
          <span>
            Already have an account? 
          </span>
          <a
            href="/login"
          >
            <span>
              Login
            </span>
          </a>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
